src/app/options/qgsgpsdeviceoptions.cpp

```diff
@@ -0,0 +1,244 @@
+/***************************************************************************
+    qgsgpsdeviceoptions.cpp
+    -------------------------
+    begin                : July 2021
+    copyright            : (C) 2021 by Nyall Dawson
+    email                : nyall dot dawson at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsgpsdeviceoptions.h"
+#include "qgsbabelgpsdevice.h"
+#include "qgssettings.h"
+#include "qgsapplication.h"
+#include "qgsbabelformatregistry.h"
+
+#include <QMessageBox>
+
+//
+// QgsGpsDeviceOptionsWidget
+//
+
+QgsGpsDeviceOptionsWidget::QgsGpsDeviceOptionsWidget( QWidget *parent )
+  : QgsOptionsPageWidget( parent )
+{
+  setupUi( this );
+
+  connect( mButtonAddDevice, &QToolButton::clicked, this, &QgsGpsDeviceOptionsWidget::addNewDevice );
+  connect( mButtonRemoveDevice, &QToolButton::clicked, this, &QgsGpsDeviceOptionsWidget::removeCurrentDevice );
+  connect( mListDevices, &QListWidget::currentItemChanged, this, &QgsGpsDeviceOptionsWidget::selectedDeviceChanged );
+
+  mDescriptionBrowser->setHtml( QStringLiteral( "<p>%1</p><ul>"
+                                "<li>%babel - %2</li>"
+                                "<li>%in - %3</li>"
+                                "<li>%out - %4</li>"
+                                "</ul>" ).arg( tr( "In the download and upload commands there can be special words that will be replaced by "
+                                    "QGIS when the commands are used. These words are:" ),
+                                    tr( "the path to GPSBabel" ),
+                                    tr( "the GPX filename when uploading or the port when downloading" ),
+                                    tr( "the port when uploading or the GPX filename when downloading" ) ) );
+
+  const QMap< QString, QgsBabelGpsDeviceFormat * > registeredDevices = QgsApplication::gpsBabelFormatRegistry()->devices();
+  for ( auto it = registeredDevices.constBegin(); it != registeredDevices.constEnd(); ++it )
+  {
+    if ( !it.value() )
+      continue;
+
+    const QString wptDownload =
```

Comment:
wptXX, rteXX, trkXX are ok for me, but waypointXX, routeXX, trackXX are may be more readable